### PR TITLE
Refactor TE fused ops integration into mixin

### PR DESCRIPTION
**Pull request by CarlosGomes98**
# What does this PR do ?
Refactor TE fused ops integration into MCore for code deduplication

Main ideas:
1. Add a mixin class that all future te fused ops integrations can use, which handles the lifecycle management for the class (e.g. registering hooks).
2. Add helper functions for building te basic ops from MCore wrapped TE modules.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
3. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
LGTM, this is a straightforward reorganization of TE ops boilerplate. This infrastructure could also be used in the [fused impl of the grouped MLP block](https://github.com/NVIDIA/Megatron-LM/blob/a2d71539dd908f3e28ace2cf7585a6a891421175/megatron/core/transformer/moe/experts.py#L842) (currently in `dev`, not yet in `main`).

File: tests/unit_tests/fusions/test_te_fused_ops.py
Comment (timmoon10) [inline tests/unit_tests/fusions/test_te_fused_ops.py:1]: Nit: It would be helpful to include a test that the `Linear`-like modules can be successfully converted to fused ops.

File: megatron/core/extensions/transformer_engine.py [hunk 1]
```diff
@@ -558,24 +570,239 @@ def forward_post_hook(module, *_) -> None:
...
+        """Return whether an adapter input exposes a possibly-nested attribute."""
+        current: Any = module
+        for part in attr_name.split("."):
+            if not hasattr(current, part):
+                return False
+            current = getattr(current, part)
+        return True
+
+    def _validate_te_ops_adapter_module(
+        module: torch.nn.Module, module_name: str, required_attrs: Tuple[str, ...]
```
Comment (timmoon10) [inline megatron/core/extensions/transformer_engine.py]: Nit:
```suggestion
        module: torch.nn.Module, module_name: str, required_attrs: Sequence[str, ...]
```

**Comment by CarlosGomes98** [comment]
/ok to test [2a02119](https://github.com/NVIDIA/Megatron-LM/pull/4630/commits/2a02119ec249abb3ccfa5605772d83dddf1eeac3)

**Comment by CarlosGomes98** [comment]
/ok to test [dcdd3a3](https://github.com/NVIDIA/Megatron-LM/pull/4630/commits/dcdd3a36ee6cdb8e6e518632df5d7ef9a99ef18f)

**Review by ilml — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by jiemingz — APPROVED** [review]

**Comment by CarlosGomes98** [comment]
https://github.com/NVIDIA/Megatron-LM/issues/5741

**Review by ilml — APPROVED** [review]

**Comment by ilml** [comment]
/ok to test [5f90ea2](https://github.com/NVIDIA/Megatron-LM/pull/4630/commits/5f90ea29d2a59ea72211994fc657e841893a119a)
